### PR TITLE
feat: Directory content type with custom attributes

### PR DESCRIPTION
**Pull request by erikgeiser**
This PR introduces the content type `dir` which can be used to add empty directories or directories with custom attributes such as owner, group or permissions. It also makes `EmptyFolders` obsolete.

Another important aspect is that it can be used to specified directories "owned" by the package when packaging RPMs as only those directories are removed when the package is uninstalled. This use case is also now well documented.

### CODE REVIEW

File: apk/apk.go [hunk 1]
```diff
@@ -399,54 +399,70 @@ func createBuilderData(info *nfpm.Info, sizep *int64) func(tw *tar.Writer) error
...
+			// fallthrough to default case`
 			fallthrough
 		default:
 			err = copyToTarAndDigest(file, tw, sizep)
 		}
 		if err != nil {
 			return err
 		}
-		created[file.Source] = true
-		created[file.Destination[1:]] = true
```
Comment (erikgeiser) [inline apk/apk.go:449]: I don't know why these two lines where there to begin with. The source only matters for symlinks but created is only to avoid creating duplicate implicit directories anyway, right?

File: apk/apk.go [hunk 2]
```diff
@@ -528,7 +532,7 @@ func pathsToCreate(dst string) []string {
 	}
 	// we don't really need to create those things in order apparently, but,
 	// it looks really weird if we don't.
-	var result []string
```
Comment (erikgeiser) [inline apk/apk.go:531]: This now matches the implementation in `deb.go`.

File: www/docs/configuration.md [hunk 3]
```diff
@@ -53,8 +53,7 @@ priority: extra
 
 # Maintainer.
 # Defaults to empty on rpm and apk
-# Leaving this field empty on 'deb' packages is deprecated, and will be removed
-# in a future release
```
Comment (erikgeiser) [inline www/docs/configuration.md:57]: This is just a drive-by fix to make it more in line with the deprecation warning.

File: deb/deb.go [hunk 4]
```diff
@@ -166,7 +165,7 @@ func (*Deb) SetPackagerDefaults(info *nfpm.Info) {
 	// if in the long run we should be more strict about this and error when
 	// not set?
 	if info.Maintainer == "" {
-		log.Println("DEPRECATION WARNING: Leaving the 'maintainer' field unset will not be allowed in a future version")
```
Comment (erikgeiser) [inline deb/deb.go:169]: We don't use `log`, elsewhere, so I changed it to print to `os.Stderr` just like the `empty_folders` deprecation warning.

**Review by djgilcrease — APPROVED** [review]
LGTM

File: apk/apk.go [hunk 5]
```diff
@@ -396,54 +396,70 @@ func createBuilderData(info *nfpm.Info, sizep *int64) func(tw *tar.Writer) error
...
-
-		// handle Files
 		return createFilesInsideTarGz(info, tw, created, sizep)
 	}
 }
 
 func createFilesInsideTarGz(info *nfpm.Info, tw *tar.Writer, created map[string]bool, sizep *int64) (err error) {
+	// create explicit directories first
+	for _, file := range info.Contents {
+		// at this point, we don't are about other types yet
```
Comment (caarlos0) [inline apk/apk.go:406]:
```suggestion
		// at this point, we don't care about other types yet
```

File: apk/apk.go [hunk 6]
```diff
@@ -396,54 +396,70 @@ func createBuilderData(info *nfpm.Info, sizep *int64) func(tw *tar.Writer) error
...
-		case "readme":
+		case "doc", "licence", "license", "readme", "config", "config|noreplace":
 			// nolint:gocritic
-			// ignoring `emptyFallthrough: remove empty case containing only fallthrough to default case`
-			fallthrough
-		case "config", "config|noreplace":
-			// nolint:gocritic
-			// ignoring `emptyFallthrough: remove empty case containing only fallthrough to default case`
+			// ignoring `emptyFallthrough: remove empty case containing only
+			// fallthrough to default case`
```
Comment (caarlos0) [inline apk/apk.go:455]: can we remove all this and let it all go to the `default`?

File: apk/apk.go (hunk 1 above)
Comment (caarlos0) [reply apk/apk.go:446]: I don't know either 🤔  but I think if tests are passing its probably something we didn't need indeed..

File: apk/apk_test.go [hunk 7]
```diff
@@ -480,3 +485,126 @@ func TestPackageSymlinks(t *testing.T) {
...
+	}
+
+	require.NoError(t, info.Validate())
+
+	var buf bytes.Buffer
+	size := int64(0)
+	err := createFilesInsideTarGz(info, tar.NewWriter(&buf), make(map[string]bool), &size)
+	require.NoError(t, err)
+
+	// for debs all implicit or explicit directories are created in the tarball
```
Comment (caarlos0) [inline apk/apk_test.go:521]: should it say for `apks` instead of debs?

File: deb/deb.go (hunk 4 above)
Comment (caarlos0) [reply deb/deb.go:174]: thinking out loud here: maybe we could allow the caller to pass a `io.Writer` or some other interface as to "where should we print the warnings" 🤔

File: deb/deb.go [hunk 8]
```diff
@@ -271,10 +267,41 @@ func createSymlinkInsideTar(file *files.Content, out *tar.Writer) error {
 
 func createFilesInsideDataTar(info *nfpm.Info, tw *tar.Writer,
 	created map[string]bool) (md5buf bytes.Buffer, instSize int64, err error) {
+	// create explicit directories first
 	for _, file := range info.Contents {
+		// at this point, we don't are about other types yet
```
Comment (caarlos0) [inline deb/deb.go:272]:
```suggestion
		// at this point, we don't care about other types yet
```

File: files/files.go [hunk 9]
```diff
@@ -125,7 +152,7 @@ func ExpandContentGlobs(contents Contents, disableGlobbing bool) (files Contents
 		var globbed map[string]string
 
 		switch f.Type {
-		case "ghost", "symlink":
+		case "ghost", "symlink", "dir":
 			// Ghost and symlink files need to be in the list, but dont glob them because they do not really exist
```
Comment (caarlos0) [inline files/files.go:156]: comment is now outdated since we added `dir` as well

**Review by caarlos0 — APPROVED** [review]
just some nitpicks, but looks good overall!

awesome work!

File: deb/deb.go (hunk 4 above)
Comment (erikgeiser) [reply deb/deb.go:174]: Another idea, a package level `io.Writer` for warnings that defaults to `os.Stderr` that can be overwritten globally. It's simple enough to add it to this PR.

**Comment by caarlos0** [comment]
awesome work as always, thanks @erikgeiser !
